### PR TITLE
change openssl api to openssl3.0

### PR DESCRIPTION
**Pull request by whh39**
Replace functions with deprecation marks in openssl3.0

### CODE REVIEW

File: core/Enclave/self_test/ecc_test.cpp [hunk 1]
```diff
@@ -304,97 +308,107 @@ EHSM_TEST_VECTOR ecc_sign_verify_test_vectors = {
...
+    BN_S = BN_bin2bn(&*S, VECTOR_LENGTH("S"), NULL);
 
-    int digestmode = atoi((test_vector["digestmode"]).c_str());
     int curve = atoi((test_vector["curve"]).c_str());
-    GET_PARAMETER(Msg);
+    int digestmode = atoi((test_vector["digestmode"]).c_str());
 
-    bool ret = false;
-    bool result = false;
+    // 创建公钥
```
Comment (yang8621) [inline core/Enclave/self_test/ecc_test.cpp:370]: no chinese

**Comment by truc0** [comment]
Could you provide more information about how to build eHSM with openssl3?

I found 2 main problem is raised when trying to build the latest commit from `whh39/fix_openssl` branch (commit [4ed37db](https://github.com/whh39/ehsm/commit/4ed37db4f626aee4818262f5e787828d3fa0d03d)) by `make` command. Build of `Dockerfile` is also failed.

### 1. `intel-sgx-ssl` submodule is not updated. 

Current version of `intel-sgx-ssl` in `ehsm` project is still an old commit, which is not compatible with openssl 3.0. It is recommended to update to latest.

However, the latest version of `intel-sgx-ssl` cannot build without modification. The main issue is `Linux/build/prepare_openssl.sh` still uses `1.1.1t` as openssl version. Change it to `*3.0*` can fix this problem and build `intel-sgx-ssl` successfully.

### 2. `libsgx_tsgx_ssl_ssl.a` missing

The `buildenv.mk` file requires `-lsgx_tsgxssl_ssl` as link library while the latest `intel-sgx-ssl` repository currently does not build.

Some references:
- [eHSM buildenv.mk](https://github.com/whh39/ehsm/blob/4ed37db4f626aee4818262f5e787828d3fa0d03d/buildenv.mk#L85)
- [Old Intel SGX SSL](https://github.com/intel/intel-sgx-ssl/blob/d5cff4a3cbac8b25c631248341c1a3ddfb404aff/Linux/build_openssl.sh#L74)
- [latest Intel SGX SSL](https://github.com/intel/intel-sgx-ssl/blob/f69f5bb3f2b347beb5c20bedf40adf56a5d371e3/Linux/prepare_openssl.sh)

By removing the `sgx_tsgxssl_ssl` from the link library list, an error as following is raised:

```
undefined reference to `PEM_read_bio_X509'
```

**Comment by yang8621** [comment]
yes. please also check in the submodule update as https://github.com/intel/ehsm/pull/305/files.

File: core/Enclave/openssl_operation.cpp [hunk 2]
```diff
@@ -966,68 +892,59 @@ sgx_status_t ecc_verify(EC_KEY *ec_key,
         mdctx = EVP_MD_CTX_new();
         if (mdctx == NULL)
         {
-            log_e("ecall ecc_verify failed to create a EVP_MD_CTX.\n");
+            log_e("ecall sm2_verify failed to create a EVP_MD_CTX.\n");
```
Comment (yang8621) [inline core/Enclave/openssl_operation.cpp:895]: ?

File: core/Enclave/openssl_operation.cpp [hunk 3]
```diff
@@ -966,68 +892,59 @@ sgx_status_t ecc_verify(EC_KEY *ec_key,
...
             ret = SGX_ERROR_OUT_OF_MEMORY;
             goto out;
         }
 
         if (EVP_MD_CTX_init(mdctx) != 1)
         {
-            log_e("ecall ecc_verify EVP_MD_CTX initialize failed.\n");
+            log_e("ecall sm2_verify EVP_MD_CTX initialize failed.\n");
```
Comment (yang8621) [inline core/Enclave/openssl_operation.cpp:902]: ?

File: core/Enclave/openssl_operation.cpp [hunk 4]
```diff
@@ -966,68 +892,59 @@ sgx_status_t ecc_verify(EC_KEY *ec_key,
...
             *result = true;
             break;
         case 0:
-            // data digest did not match the original data or the signature had an invalid form
+            // tbs did not match the original data or the signature had an invalid form
             *result = false;
             break;
         default:
-            log_e("ecall ecc_verify ECDSA_verify failed.\n");
+            log_e("ecall sm2_verify EVP_DigestVerifyFinal failed.\n");
```
Comment (yang8621) [inline core/Enclave/openssl_operation.cpp:930]: ?

File: core/Enclave/enclave_hsm.cpp [hunk 5]
```diff
@@ -664,12 +664,7 @@ sgx_status_t enclave_generate_datakey(ehsm_keyblob_t *cmk, size_t cmk_size,
             ciphertext->datalen = plaintext->datalen + EH_AES_GCM_IV_SIZE + EH_AES_GCM_MAC_SIZE;
             return SGX_SUCCESS;
         case EH_SM4_CBC:
-            if (plaintext->datalen % 16 != 0)
-            {
-                ciphertext->datalen = (plaintext->datalen / 16 + 1) * 16 + SGX_SM4_IV_SIZE;
-                return SGX_SUCCESS;
-            }
-            ciphertext->datalen = plaintext->datalen + SGX_SM4_IV_SIZE;
+            ciphertext->datalen = (plaintext->datalen / 16 + 1) * 16 + SGX_SM4_IV_SIZE;
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:667]: add comments, why do this change? bug fix?

File: core/Enclave/enclave_hsm.cpp [hunk 6]
```diff
@@ -771,37 +765,38 @@ sgx_status_t enclave_export_datakey(ehsm_keyblob_t *cmk, size_t cmk_size,
...
-    tmp_datakey->datalen = tmp_datakey_tmp.datalen;
+    tmp_datakey->datalen = tmp_datakey_size;
+    tmp_datakey_size = APPEND_SIZE_TO_DATA_T(tmp_datakey_size);
     // decrypt olddatakey using cmk
     switch (cmk->metadata.keyspec)
     {
     case EH_AES_GCM_128:
     case EH_AES_GCM_192:
     case EH_AES_GCM_256:
+    case EH_SM4_CTR:
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:794]: newly added algo? if so, please move it into a new commit.

File: core/Enclave/key_operation.cpp [hunk 7]
```diff
@@ -1294,12 +1259,14 @@ sgx_status_t ehsm_sm2_sign(const ehsm_keyblob_t *cmk,
...
+    uint8_t *data = NULL;
+    OSSL_DECODER_CTX *dctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    size_t dataLen = 0;
 
     // Get padding mode and digest mode
     const EVP_MD *digest = getDigestMode(digest_mode);
-    if (digest == NULL)
+    if (digest != EVP_sm3())
         digest = EVP_sm3();
```
[the remark this answers is not in the document]
Comment (whh39) [reply core/Enclave/key_operation.cpp:1270]: to add a default digest for sm2, ec sign/verify has same code

**Review by syan10 — APPROVED** [review]

File: core/Enclave/key_operation.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (yang8621) [reply core/Enclave/key_operation.cpp:1270]: if (digest == NULL)
        digest = EVP_sm3();
以前这样才是default吧

[the remark this answers is not in the document]
Comment (syan10) [reply core/Enclave/key_operation.cpp:1270]: 这个digest_mode是用户带进来的参数吗？  是必填的？

你这么改是担心他回传其他的digest mode？ 比如sha256之类的？  如果这样的话， 我觉得直接return 一个error是不是更好？ 对于sm2只能是sm3的digest

File: core/Enclave/key_operation.cpp [hunk 8]
```diff
@@ -1365,12 +1332,14 @@ sgx_status_t ehsm_sm2_verify(const ehsm_keyblob_t *cmk,
...
-    EC_KEY *ec_key = NULL;
+    uint8_t *data = NULL;
+    OSSL_DECODER_CTX *dctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    size_t dataLen = 0;
 
     // Get padding mode and digest mode
     const EVP_MD *digest = getDigestMode(digest_mode);
-    if (digest == NULL)
+    if (digest != EVP_sm3())
```
Comment (yang8621) [inline core/Enclave/key_operation.cpp:1342]: and here

**Comment by yang8621** [comment]
add third_party/intel-sgx-ssl changes

**Review by yang8621 — APPROVED** [review]

**Pull request merged**